### ISSUE TITLE
Move the Coding Style Guidelines to the code_style checklist

### ISSUE THREAD

**Issue by samreid** [phetsims/phet-info#48]
Currently the Coding Style Guidelines has this section:

![image](https://user-images.githubusercontent.com/679486/27644850-d116c5cc-5be1-11e7-993f-ca3dee5f7636.png)

This refers to an external document that lists 23 items to be checked, 18 of which are not covered by ESLint.  Since these are not part of the checklist, we have no easy way to visualize if these have been pass or fail for code reviews.  It seems odd that our code review checklist has 63 items plus a link to 18 other items that don't have a checklist.  It gives the impression that these 18 items are less important or don't need to be tracked, when in fact, many of them are as important or more important than the other 63 items.

In https://github.com/phetsims/phet-info/issues/36#issuecomment-261336582 we discussed moving the "Coding Style Guidelines" to the checklist and decided not to, but no explanation is given.  The only reasons I can think of not to move the checklist are:
1. it would make the checklist longer
2. There would need to be a link to the style guidelines in the PhET Development Overview
3. It would take time to move the checklist.

These do not seem like compelling reasons for separating these items. In my opinion, we should move these 18 code style items into the main sim_review checklist.

Tagging @jonathanolson @jbphet @jonathanolson @jessegreenberg @pixelzoom @zepumph @Denz1994 @phetsims/martin  for discussion.

EDIT: Also tagging @ariel-phet for discussion

**Comment by samreid** [comment]
> There would need to be a link to the style guidelines in the PhET Development Overview

We should add that anyways.

**Comment by samreid** [comment]
I've committed a rough draft to this branch: https://github.com/phetsims/phet-info/blob/style-checklist/checklists/code_review_checklist.md

**Comment by pixelzoom** [comment]
My recollection is that we decided to leave these items in the coding-style document because:

 (1) They are applicable to anyone writing sim code, and those people are unlikely to be using the code-review checklist. 

(2) Duplicating in both places is not desirable.

(3) PhET developers should be highly aware of PhET code guidelines, so 1 item in the review checklist should be sufficient. 

So my vote is to leave as is.

**Comment by pixelzoom** [comment]
@samreid said:
> In my opinion, we should move these 18 code style items into the main sim_review checklist.
My recollection is that we decided to leave these items in the coding-style document because:

I disagree.

 (1) They are applicable to anyone writing sim code, and external developers are unlikely to be using the code-review checklist. 

(2) Duplicating in both places is not desirable.

(3) PhET developers should be highly aware of PhET code guidelines, so 1 item in the review checklist should be sufficient. 

So my vote is to leave as is.

**Comment by samreid** [comment]
> (1) They are applicable to anyone writing sim code, and those people are unlikely to be using the code-review checklist.

Why would we want people writing sim code to not be using the code review checklist?

> (2) Duplicating in both places is not desirable.

We will have one copy in the code review checklist.

> (3) PhET developers should be highly aware of PhET code guidelines, so 1 item in the review checklist should be sufficient.

We need a checklist for the "style" items for the same reasons we need a checklist for the other items, so we can be mindful of them, not forget them and make sure they are addressed.

**Comment by pixelzoom** [comment]
> Why would we want people writing sim code to not be using the code review checklist?

Not what I said.  I said that external developers were unlikely to find the code review checklist.

> We will have one copy in the code review checklist.

Currently not the case.  There is currently no reference to the review checklist, and your above-comitted draft now duplicates the items across 2 documents. 

> We need a checklist for the "style" items for the same reasons we need a checklist for the other items, so we can be mindful of them, not forget them and make sure they are addressed.

If you need a check list for things that you should be doing every day, then there's a problem that putting them in a check list is not going to solve.

Imo, this issue is busy work.

**Comment by samreid** [comment]
> I said that external developers were unlikely to find the code review checklist.

The code review checklist is an important document that sim developers should be aware of.  We should link to it or feature it in such a way that it can be easily discovered.

> There is currently no reference to the review checklist, and your above-comitted draft now duplicates the items across 2 documents.

I'm working in a branch and leaving the google doc alone until we have a quorum.  Once this work is approved, we will eliminate the copy from the google doc and add a link to the code review checklist.

> If you need a check list for things that you should be doing every day, then there's a problem that putting them in a check list is not going to solve.

Code style issues regarding read-only and short circuiting have been up-for-discussion lately and putting them in the code review checklist seems like a good way to make sure we are on the same page.  If any of the items is so obvious that we don't want it checked in a code review, then perhaps we should eliminate it from the code review checklist.  For instance, "Names (types, variables, properties,...) should be sufficiently descriptive and specific, and should avoid non-standard abbreviations" seems like a great thing to double check during review.  But maybe we no longer need an item for "All JavaScript scripts should be modularized using requirejs"???

**Comment by samreid** [comment]
Since I've only heard from @pixelzoom, I'll move this to Developer Meeting agenda.

**Comment by samreid** [comment]
@zepumph said this could be a separate checklist

@ariel-phet said we could remove the "all script should be modularized with requirejs"

**Comment by samreid** [comment]
Consensus: leave them in checklist.  Move it higher in the list.  Remove unnecessary things like "requirejs modularization".

**Comment by samreid** [comment]
Merged to master, I'll continue on the above recommendations.

**Comment by pixelzoom** [comment]
... and replace code-style in PhET Development Overview with a link to the code-review checklist.

**Comment by samreid** [comment]
I updated the Development Overview style guidelines to point to the checklist: https://docs.google.com/document/d/1Ys1EiwnqQGYuzGOcQSr4uXDes35mF1v1XhMZIl10nk8/edit#heading=h.1oxr3ptyo50w

**Comment by samreid** [comment]
I also removed the requirejs bit, fixed the formatting (there were bad `<` characters) and moved it higher in the list. I also cleaned up one of the poorer examples. 

@pixelzoom would you like to take a look?  I wasn't sure exactly how high to move the "Coding Conventions" in the list.

**Comment by pixelzoom** [comment]
There is now a circular reference.  In the checklist:

> - [ ] Are documentation conventions followed, as described in the “Coding Style Guidelines” section of the PhET Development Overview?

One approach would be to move items in the "Coding Conventions" section of the checklist that are related to documentation to the "Documentation" section of the checklist.  But that looks like a can of worms.   Not sure how to proceed...

**Comment by zepumph** [comment]
One thought is that "Coding Conventions" means how you code, which includes documentation, so maybe we could remove the aforementioned bullet that creates the circular reference, and just go from there. It will still be covered in the conventions section of the review.

**Comment by pixelzoom** [comment]
+1 for keeping coding-related documentation stuff in "Coding Conventions".   

As for the other items that are currently under the "Documentation" heading...

There 3 could probably move to the "Repository Structure" section:

> - [ ] Does model.md adequately describe the model, in terms appropriate for teachers?
> - [ ] Does implementation-notes.md adequately describe the implementation, with an overview that will be useful to future maintainers?
> - [ ] Are sim-specific query parameters (if any) identified and documented in one .js file in js/common/ or js/ (if there is no common/)? The .js file should be named {{REPO}}QueryParameters, for example ArithmeticQueryParameters for the aritmetic repository.

This item should probably move to "Coding Conventions":

> - [ ] Do the @author annotations seem correct?

... then the "Documentation" section could be deleted.

**Comment by pixelzoom** [comment]
I did what I described above - moved Documentation items to other sections, then deleted Documentation section.

@samreid please review.

**Comment by samreid** [comment]
I reviewed all of the deltas and all seems well, thanks!  Oh, one spelling fix is on the way.

**Comment by samreid** [comment]
Also, I realized there is no checkbox for "every property method should be documented" (though it does say the visibility annotation should be documented).  Should we add a checkbox for that?

**Comment by pixelzoom** [comment]
> ... there is no checkbox for "every property method should be documented"...

I think it's fine to _not_ document ubiquitous methods like `reset` and `dispose`, or obvious methods like `setWidth` and `getWidth`.   Otherwise I'm ambivalent about the need for this checklist item.

**Comment by jonathanolson** [comment]
> I think it's fine to not document ubiquitous methods like reset and dispose

On the other hand, presumably we'd want them to be documented if we are ever generating documentation for phet-io use?

**Comment by samreid** [comment]
I added this note:

`- [ ] Every class, method and property should be documented.`

Closing for now!  Anyone reopen if you wish.

**Comment by pixelzoom** [comment]
I changed `class` to `type`.  JavaScript does not have classes.

**Comment by samreid** [comment]
> JavaScript does not have classes.

It will when we drop IE support :) . https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes
